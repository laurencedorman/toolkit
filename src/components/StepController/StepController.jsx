/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class StepController extends PureComponent {
  static defaultProps = { getActiveStep: undefined };

  constructor(props) {
    super(props);
    this.state = {
      activeStep: 0,
      totalStep: undefined,
      direction: 'up',
      lastStep: false,
    };
  }

  componentDidMount() {
    const { getActiveStep } = this.props;
    if (getActiveStep) this.setActiveStep(getActiveStep);
  }

  setActiveStep = activeStep => this.setState({ activeStep });

  setTotalStep = totalStep => this.setState({ totalStep });

  prevStep = () => {
    const { totalStep } = this.state;

    this.setState(prevState => ({
      direction: 'down',
      activeStep: prevState.activeStep === 0 ? 0 : prevState.activeStep - 1,
      lastStep: prevState.activeStep !== totalStep - 1 && false,
    }));
  };

  nextStep = () => {
    const { totalStep } = this.state;

    this.setState(prevState => ({
      direction: 'up',
      activeStep:
        prevState.activeStep === totalStep - 1
          ? totalStep - 1
          : prevState.activeStep + 1,
      lastStep: prevState.activeStep === totalStep - 1 && true,
    }));
  };

  render() {
    const { children } = this.props;
    const { activeStep, direction, lastStep, totalStep } = this.state;

    return children({
      prevStep: this.prevStep,
      nextStep: this.nextStep,
      setTotalStep: this.setTotalStep,
      activeStep,
      direction,
      lastStep,
      totalStep,
    });
  }
}

StepController.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  getActiveStep: PropTypes.oneOfType([PropTypes.number, PropTypes.func]),
};
