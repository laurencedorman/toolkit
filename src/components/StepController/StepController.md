StepperControl with RenderChild pattern :

<br/>
<br/>
  
State to Props | Type | Description
-------------- | ---- | -----------
prevStep | func | Decrement (-1)
nextStep |func | Increment (+1)
setTotalStep | func | Set total children of controlled target
activeStep | number | index odd active step
direction | string | 'up' or 'down
lastStep | boolean | true when last step is active
totalStep | number | return total step (use setTotalStep before to set number)

```js
import Button from '../Button';
import Card from '../Card';
import Wrapper from '../Wrapper';

<StepController>
  {({ prevStep, nextStep, activeStep, setTotalStep }) => (
    <Wrapper
      direction="row"
      style={{ alignItems: 'center', justifyContent: 'space-around' }}
    >
      <Button onClick={prevStep}>previous step</Button>
      <Card>{activeStep}</Card>
      <Button onClick={nextStep}>next step</Button>
    </Wrapper>
  )}
</StepController>;
```
