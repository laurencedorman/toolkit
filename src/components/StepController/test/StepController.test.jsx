/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow } from 'enzyme';
import StepController from '../StepController';

describe('StepController', () => {
  const renderMock = jest.fn();
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <StepController getActiveStep={renderMock}>
        {({ props }) => props}
      </StepController>
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state on mount', done => {
    setImmediate(() => {
      expect(wrapper.state('activeStep')).toEqual(renderMock);
      done();
    }, 0);
  });

  it('should set direction state', () => {
    wrapper.setState({ direction: 'down' });
    expect(wrapper.state('direction')).toEqual('down');
  });

  it('should set lastStep state', () => {
    wrapper.setState({ lastStep: '5' });
    expect(wrapper.state('lastStep')).toEqual('5');
  });
});
