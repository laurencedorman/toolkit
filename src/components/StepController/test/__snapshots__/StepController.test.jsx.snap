// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepController should match snapshot 1`] = `""`;
